### ISSUE TITLE
Feature: Support for Mariadb

### ISSUE THREAD

**Issue by landbaychrisburrell** [kviklet/kviklet#157]
Hi

I've been trying to get this running locally on our internal stack. We use Mariadb, which has strong compatibility, but not 100%, with MariaDB. The first issue was trivial to solve. I'm happy to contemplate putting a PR together for resolving the second - but would need a pointer as to how you want to support it.

First issue - (see patch) - the liquibase doc also suggests changeColumnDataType is missing
[Allowing_mariadb_connection_to_start_server_.patch](https://github.com/user-attachments/files/16140000/Allowing_mariadb_connection_to_start_server_.patch)

The second issue is (I guess?) likely to do with the fact Mariadb has got a json alias (to a varchar) but not a proper type. I haven't come across the "::json?" notation previously in the converters previously. I suspect this will be one of those that is different dependant on the backend - any pointers as to how you would like to fix this?

```Resolved [org.springframework.dao.InvalidDataAccessResourceUsageException: could not execute statement [(conn=347) You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '::json,'root','asdasd')' at line 1] [insert into connection (additional_jdbc_options,authentication_type,connection_type,database_name,datasource_type,description,display_name,hostname,max_executions,password,port,review_config,username,id) values (?,?,?,?,?,?,?,?,?,?,?,?::json,?,?)]; SQL [insert into connection (additional_jdbc_options,authentication_type,connection_type,database_name,datasource_type,description,display_name,hostname,max_executions,password,port,review_config,username,id) values (?,?,?,?,?,?,?,?,?,?,?,?::json,?,?)]]```

For ease of replication, I'm running MariaDb version 11.1.3, with the rest as defaults.

`

**Comment by Askir** [comment]
Hey, I haven't updated that Readme section in ages and only tried it once aswell, I should have probably removed the reference to mysql/mariadb as the kviklet DB instead of raising wrong hopes.

Is there a particular reason you cannot run postgres? I think I could try fix these problems. Metabase for example makes it work somehow I believe, but of course supporting two database types instead of one is a bit of a maintenance overhead, so I'm wondering if that is really worth it 😅

**Comment by landbaychrisburrell** [comment]
Hi - thanks for quick response.

All of our databases are on Mariadb, hosted on AWS. We have approximately 60 database schemas across 2 Mariadb servers, both Mariadb, multi-AZ in AWS. We have an internal dev team of around 25 engineers who are all used to maintaining and building against Mariadb. That includes upgrading Mariadb regularly, etc.

If we really had to, we could spin up a small postgres instance at low cost, but at the moment, my mindset is more in the space of reducing the infrastructure surface to keep our devs doing Java build, rather than having to update 'another db' instance regularly.

Great product idea btw - in March, we almost built something exactly like kviklet but got delayed a bit, so when we saw kviklet we thought we'd try it as it's an exact fit and further than we would have taken it.

**Comment by landbaychrisburrell** [comment]
Thanks - AWS's auto-updates are mostly for minor updates (which we have turned on) so would still have to deal with major upgrades.

(Look forward to perhaps one day seeing the proxy for MySQL/Mariadb feature. One other feature to consider would be for IAM roles, which we were going to build - similar to ConsoleMe by Netflix but a much lighter way)

**Comment by Askir** [comment]
Building the Proxy for postgres was super fun but honestly my confidence in it is rather low because:
- I have only tested it with one specific postgres version
- I have only used it with my db client of choice (Datagrip)
- It runs as an extra thread on the Kviklet container almost like another separate process, depending on what you do i'm sure you could exhaust recourses quickly
- It doesn't support SSL encryption between the user and Kviklet atm

The last one might be fixable, but fixing the other three confidently simply requires quite a bit of effort. Doing this for multiple databases... I think that's definitely a full-time job if not two. There is tools like Teleport and StrongDM out there, that do this btw but they have hundreds of engineers and the tools are correspondingly expensive.

So, unless I figure out a way to do Kviklet full-time or other contributors provide the code more proxy support is unlikely to happen I think.

___

Regarding IAM roles, do you mean temporary access to specific resources managed in the same way as DB access is managed? E.g. a user requesting temporary access to a specific role to write/read from an SQS queue and only getting it after approval?

I have actually never heard of ConsoleMe but I took a look and it looks to solve a variety of problems, some of which I think are better solved via good IaC management. But something like the mentioned use-case, I'd be happy to bring to Kviklet eventually.

**Comment by Askir** [comment]
I spent some time today on figuring out what's going on but don't have an immediate solution yet.
I am making use of postgres' native json columns in multiple places. To make Kviklet compatible with multiple database engines for storing state I'd have to convert those back to text columns.

This means writing a migration and changing the code so that it can deal with text columns, but it also means no more json based querying. Which tbf I only use for debugging purposes so far but is kinda handy sometimes.

Now I also don't know if this support of multiple DBs really something I should strive for.
I looked at some other selfhosted tools. For reference: Gitlab and Langfuse both only support PostgreSQL for self hosting. Metabase  supports mariadb, mysql and postgres. Sonarqube support MS sql server, oracle and postgres. Grafana supports mysql and postgres.

Postgres is definitely the common denominator here 😅 Gitlab has a post on why they ended supporting MySQL as well: https://about.gitlab.com/blog/2019/06/27/removing-mysql-support/

So long story short -> I'm leaning towards postgres only so far.

**Comment by Askir** [comment]
I removed mentions of mysql from the readme.
> We've seen a steady migration to PostgreSQL and recently had less than 1200 GitLab instances reporting their usage of MySQL. At the time there were 110,000 instances using PostgreSQL.

[Just in general seems like the industry is shifting to postgres as the clear favorite](https://tamerc.com/posts/ask-hn-who-is-hiring/#which-databases-are-used)

Sorry to disappoint on this front. Happy to help adapt Kviklet on other features for your use cases though!
